Sources/_StringProcessing/LiteralPrinter.swift

```diff
@@ -0,0 +1,527 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2021-2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+@_implementationOnly import _RegexParser
+
+@available(macOS 9999, *) // TODO: 5.9?
+extension Regex {
+  /// The literal pattern for this regex.
+  ///
+  /// This is non-`nil` when used on a regex that can be represented as a
+  /// string. The literal pattern may be different from the literal or string
+  /// that was used to create the regex. If this regex includes components
+  /// that cannot be represented in a regex literal, such as a capture transform
+  /// or a custom parser that conforms to the `CustomConsumingRegexComponent`
+  /// protocol, this property is `nil`.
+  ///
+  /// The value of this property may change between different releases of Swift.
+  // TODO: SPI
+  public var _literalPattern: String? {
+    var gen = LiteralPrinter(options: MatchingOptions())
+    gen.outputNode(self.program.tree.root)
+    return gen.canonicalLiteralString
+  }
+}
+
+enum PatternSegment {
+  case converted(String)
+  case inconvertible(DSLTree.Node)
+  
+  var string: String? {
+    switch self {
+    case let .converted(str):
+      return str
+    case .inconvertible:
+      return nil
+    }
+  }
+}
+
+fileprivate struct LiteralPrinter {
```

Comment:
Should this replace `printAsCanonical`, or merge them?

Reply:
This doesn't necessarily need to _replace_ `printAsCanonical`, since they work on different things. If we have the AST -> DSL conversion, though, then this should be able to print anything, and we wouldn't need the existing AST printer any more. Not sure if we need the headache of keeping both in sync